kart/dataset2_paths.py

```diff
@@ -0,0 +1,174 @@
+from .exceptions import InvalidOperation
+from .serialise_util import (
+    msg_pack,
+    b64encode_str,
+    b64hash,
+    hexhash,
+)
+
+
+class PathEncoder:
+    """
+    A system for encoding a feature's primary key to a particular path.
+    Originally hex-hash[0:2]/hex-hash[2:4]/base64-encoded-message-packed-primary-key -
+    - that is, 2 levels, and a branch factor of 256 at each level.
+    But, this system proved costly for massive repos and the resulting massive trees.
+    Now we prefer 4 levels, a branch factor of 64, and clustering similar PKs where possible.
```

Comment:
Rather than start out by describing the legacy system, I'd put the current system first and put the legacy one in an epilogue

```suggestion
    A dataset with a single integer primary key will get encoded with 4 tree levels
    and a branch factor of 64 at each level. Sequential PKs tend to end up in the
    same tree.
    Anything else (multiple PK fields, strings, etc) gets the PKs values hashed first,
    before encoding into a similar 64-branch 4-level structure.

    Before 0.10, Kart used a two-level, 256-branch structure and hashed *all* PKs first.
    However, this system proved costly for massive repos (roughly, repos with greater
    than 16 milllion features), resulting in significant repository bloat.
    Existing repos generated before this change continue to use the old system, and
    Kart will continue to be compatible with those repos.

```

Reply:
Done